### ISSUE TITLE
[Feature request] generate cli parser from an object definition

### ISSUE THREAD

**Issue by yglukhov** [c-blake/cligen#2]
While playing with cligen I thought that another convenient feature would be generating the parser from an object, or from complex objects, accepted as args from the main function. Example:
```nim
type CliOptions = object # or ref object. or tuple...
  someArg1: string
  someArg2: bool

proc mainProc(options: CliOptions): int = discard
```
```
myTool --someArg1=hi --someArg2
```
This way it is much easier to pass cli options from within mainProc to other procs. The idea is pretty raw and needs more thought regarding positional args, etc.

**Comment by c-blake** [comment]
I would steer away from a _mandate_ for this (at least under the same macro names dispatchGen/dispatch) because I think the current "loose coupling" is a good property. I don't think you're really asking for a mandate, though.

In terms of refining the idea, it can become a bit thorny/require thought for how to handle "recursive objects" - fields within our object like ``someArg3: anotherObjectType``.  It could also be that however we figure out how to deal with that latter case would then also apply to the top-level case of an object in the ``mainProc`` parameter list.  There are also namespace questions in how to map/export object fields to the CLI.  E.g., ``procMain`` could also have a ``someArg1`` parameter as well as the one within the object.  Should the CLI option be ``--CliOptions-someArg1`` and the top-level one be simply ``--someArg1``?

I should also observe that, at present, you can write your own ``argParse`` template that could take a single option argument and bind that to the fields of an object with some custom "internal" syntax like --options ":someArg1=val1 :someArg2=val2".  I admit this isn't the nicest thing to use from the command-line, but it would achieve initializing object/sub-object/etc. data from command arguments.  Anyway, I'm open to the idea, but I am not quite sure how to make it all both "nice" and "general".

**Comment by c-blake** [comment]
One possible direction might be to try to try to set up a mapping between "arbitrary Nim L-value-ish expressions" and "long option names" with some new macro or new macro argument along the lines of ``{ "someArg1" : options.someArg1, "someArg2" : newObj.complexField.arg2, ... }``.  In that approach, though, it might be a bit tricky with that approach to generate the "allocation code" in the dispatcher to create the required  container object hierarchy.

**Comment by pb-cdunn** [comment]
I had the same idea, mainly because I want to be able to show all the inputs that are *actually* received. But we do not need a change to **cligen**. Just use a wrapper:
```nim
proc mainProc(options: CliOptions): int =
  echo "Your options were:", repr(options)

proc wrapper(someArg1=1, someArg2="two"): int =
  options = CliOptions(someArg1, someArg2)
  mainProc(options)

cligen.dispatch(wrapper)
```
That has exactly as much redundancy as any object "init" proc. The main drawback: `wrapper` cannot be used as the "init" proc because it also dispatches.

So the one addition I might like to **cligen** would be a non-dispatching parser that would discern the help-message from a real "init" proc, which returns an object. I would then dispatch it myself.

**Comment by c-blake** [comment]
Maybe what is simplest here is not "not dispatching", but using the ``dispatchGen`` macro instead of ``dispatch``.  The latter calls ``quit(dispatchFoo(..))`` making the program quit/live as long as the dispatched proc.  The former merely defines a ``dispatchFoo`` you can call manually.  It's true that it's return value is an integer exit code, possibly indicating invalid user inputs.  However,the proc which gets called/dispatched to can just initialize fields in some a global (or nested local) scope object variable.  I think you could do something like this with the existing code.

There might also be some entirely "other way" to go about this sort of thing where the generated proc does not try to return an exit code or even print help messages/error messages/make a whole "program" for the user.  Instead, maybe the wrapped proc has a result type which is complex (like the init* type things) and the generated proc tries to assign to fields of said "result" variable instead of assigning local variables patterned after the formal proc parameter list, and maybe the generated proc assigns to a  help: var string, and a error: int that callers can catch and use as they like.  There might need to be name similarity (or a mapping ``{}``) between the proc-parameters and the corresponding fields of the result variable type.  This way is probably not just tweaking the existing stuff, but writing another set of similar macros probably different enough to not warrant. doing it all in one family of macros.  Or maybe there's a refactoring/layering that could provide both sorts of generators.

**Comment by pb-cdunn** [comment]
Could you describe more concretely how I would use `dispatchGen` with `CliOptions` above?

**Comment by c-blake** [comment]
Well, I guess what I was thinking was something like this:
```Nim
proc mainProc(options: CliOptions): int =
  echo "Your options were:", repr(options)

var options: CliOptions
proc init(someArg1=1, someArg2="two") =
  options = CliOptions(someArg1, someArg2)
 
dispatchGen(init, cmdName="whatever")
discard dispatchInit()          # init an options global from command params
quit(mainProc(options))
```
Looking at it written out, I guess one might want ``init`` to be more general, such as returning a ``CliOptions`` object or taking a ``var`` one or something.  The idea still works if you can make a nil default (or some similar default value argument protocol) make ``init`` instead initalize some global variable like ``options`` for the later quit(mainProc()) to use, though.  You might need to add a ``suppress = @[ "myarg" ]`` to your ``dispatchGen`` for that new parameter to keep any user from trying to enter something for that.

Does that make sense to you?  Basically, there was always a non-dispatching parser in ``cligen``.  It's just a little more work to use it.

**Comment by pb-cdunn** [comment]
Ok. I see what you're saying. Because `dispatchInit()` can only return an integer, we need a global to store the options. (I'd rather not deal with argument suppression.)

I didn't realize that the `dispatchGen()` macro causes `dispatchInit()` to exist. Interesting. Thanks.

**Comment by c-blake** [comment]
Yeah. Arg suppression/the generality of your ``init`` is totally up to you.  I was just mentioning that as one way to go.  It would probably be more complicated than is worthwhile to restructure things for ``dispatchInit`` to return non-``int``.

``dispatchGen`` is really the meat of the system.  ``dispatch`` itself is a _paper-thin_ wrapper that literally just calls ``dispatchGen(...); quit(dispatchXyz())``.  ``dispatchMulti`` is more involved, but only a bit more..a loop to call ``dispatchGen`` N times and then a mini dispatcher to route dispatch/help generation to subcommands.  Maybe I should have a ``HACKING.md`` file or something, too.

**Comment by c-blake** [comment]
Resolved by https://github.com/c-blake/cligen/commit/07e0f319adf13ce3eca3b69b0acf6d0e63de5812
